File: ada/ada2022/TypeScript/Symbol.ts
import { TypeClassification } from './TypeClassification';

export class Symbol {
    name: string;
    classification: Set<TypeClassification>;
    members: Map<string, Symbol> = new Map();
    parent: Symbol | null = null;
    predefined: boolean = false;
    isComposite: boolean = false;
    definedFile: string = "";
    definedLine: number = 0;
    definedColumn: number = 0;

    constructor(name: string, classification: Set<TypeClassification>) {
        this.name = name;
        this.classification = classification;
    }

    toString(): string {
        let result = this.name;
        result += " (with classification " + Array.from(this.classification).join(", ") + ")";
        if (this.isComposite) {
            result += " [composite]";
        }
        if (this.definedFile !== "") {
            result += " defined at " + this.definedFile + ":" + this.definedLine + ":" + this.definedColumn;
        }
        return result;
    }
}

File: ada/ada2022/Dart/Symbol.dart
import 'TypeClassification.dart';

class Symbol {
  String name = "";
  Set<TypeClassification> classification = {};
  Map<String, Symbol> members = {};
  Symbol? parent;
  bool predefined = false;
  bool isComposite = false;
  String definedFile = "";
  int definedLine = 0;
  int definedColumn = 0;

  @override
  String toString() {
    var result = name;
    result += " (with classification ${classification.map((c) => c.name).join(', ')})";
    if (isComposite) result += " [composite]";
    if (definedFile.isNotEmpty) result += " defined at $definedFile:$definedLine:$definedColumn";
    return result;
  }
}
